File: core/wire/messages.go
package wire

import (
	"encoding/json"
	"fmt"
)

// Protocol constants shared with the Swift implementation (docs/protocol.md).
const (
	Version     = "0.2"
	ServiceType = "_cndt-app._tcp"

	// Capability identifiers (feature-flag strings; peers use the intersection).
	CapFile         = "file"
	CapClipboard    = "clipboard"
	CapInputInject  = "input-inject"
	CapMediaTarget  = "media-target"
	CapScreenSource = "screen-source"
	CapScreenView   = "screen-view"
	CapNotifySource = "notify-source"
	CapNotifyShow   = "notify-show"

	DefaultChunkSize uint32 = 512 * 1024
)

// MessageType is the envelope `type` discriminator.
type MessageType string

const (
	TypeHello         MessageType = "HELLO"
	TypeHelloAck      MessageType = "HELLO_ACK"
	TypePing          MessageType = "PING"
	TypePong          MessageType = "PONG"
	TypeClipboardPush MessageType = "CLIPBOARD_PUSH"
	TypeFileOffer     MessageType = "FILE_OFFER"
	TypeFileAccept    MessageType = "FILE_ACCEPT"
	TypeFileReject    MessageType = "FILE_REJECT"
	TypeFileAck       MessageType = "FILE_ACK"
	TypePairRequest   MessageType = "PAIR_REQUEST"
	TypePairResponse  MessageType = "PAIR_RESPONSE"
	TypePairConfirm   MessageType = "PAIR_CONFIRM"
	TypePairReject    MessageType = "PAIR_REJECT"
	TypeBulkAttach    MessageType = "BULK_ATTACH"
	TypeInputRequest  MessageType = "INPUT_REQUEST"
	TypeInputStatus   MessageType = "INPUT_STATUS"
	TypeInputEvent    MessageType = "INPUT_EVENT"
	TypeInputAttach   MessageType = "INPUT_ATTACH"
	TypeMediaControl  MessageType = "MEDIA_CONTROL"
	TypeScreenRequest MessageType = "SCREEN_REQUEST"
	TypeScreenOffer   MessageType = "SCREEN_OFFER"
	TypeScreenReject  MessageType = "SCREEN_REJECT"
	TypeScreenAttach  MessageType = "SCREEN_ATTACH"
	TypeScreenAck     MessageType = "SCREEN_ACK"
	TypeScreenEnd     MessageType = "SCREEN_END"
	// Phase 4-5: notifications.
	TypeNotification MessageType = "NOTIFICATION"
)

// Envelope carries every control message. Frozen shape (spec §6):
// version, type, session_id, seq, payload.
type Envelope struct {
	Payload   json.RawMessage `json:"payload"`
	Seq       uint64          `json:"seq"`
	SessionID string          `json:"session_id"`
	Type      string          `json:"type"`
	Version   string          `json:"version"`
}

// --- Message bodies. Fields are declared in the order that keeps hand-written
// struct marshaling readable; Canonical() re-sorts keys, so declaration order
// does not affect the wire bytes. ---

type DeviceClass string

type HelloBody struct {
	Identity      string   `json:"identity"`
	Name          string   `json:"name"`
	DeviceClass   string   `json:"device_class"`
	AppVersion    string   `json:"app_version"`
	Pubkey        []byte   `json:"pubkey"`
	Capabilities  []string `json:"capabilities"`
	PlatformWalls []string `json:"platform_walls"`
	ListenPort    *uint16  `json:"listen_port,omitempty"`
}

type PingBody struct {
	Nonce string `json:"nonce"`
	T     uint64 `json:"t"`
}

type ClipboardPushBody struct {
	Mime string `json:"mime"`
	Data []byte `json:"data"`
}

type FileOfferBody struct {
	FileID     string `json:"file_id"`
	Name       string `json:"name"`
	Size       uint64 `json:"size"`
	Mime       string `json:"mime"`
	SHA256     string `json:"sha256"`
	ChunkSize  uint32 `json:"chunk_size"`
	ChunkCount uint64 `json:"chunk_count"`
}

type FileAcceptBody struct {
	FileID          string `json:"file_id"`
	ResumeFromChunk uint64 `json:"resume_from_chunk"`
	BulkToken       string `json:"bulk_token"`
}

type FileRejectBody struct {
	FileID string `json:"file_id"`
	Reason string `json:"reason"`
}

type FileAckBody struct {
	FileID       string `json:"file_id"`
	Status       string `json:"status"`
	AckedThrough uint64 `json:"acked_through"`
	Message      string `json:"message,omitempty"`
}

type PairBody struct {
	Identity        string `json:"identity"`
	Name            string `json:"name"`
	DeviceClass     string `json:"device_class"`
	Pubkey          []byte `json:"pubkey"`
	TLSPubkeySHA256 string `json:"tls_pubkey_sha256"`
	BindingSig      []byte `json:"binding_sig"`
}

type PairRejectBody struct {
	Reason string `json:"reason"`
}

type BulkAttachBody struct {
	FileID    string `json:"file_id"`
	BulkToken string `json:"bulk_token"`
}

type NotificationBody struct {
	AppName string   `json:"app_name"`
	Title   string   `json:"title"`
	Body    string   `json:"body"`
	ID      string   `json:"id"`
	Actions []string `json:"actions,omitempty"`
}

// --- Phase 2: remote input + media. Optional fields are pointers so a nil
// (absent) is distinct from a present zero, matching Swift's Optional encoding. ---

type InputEventBody struct {
	Kind       string   `json:"kind"`
	Dx         *float64 `json:"dx,omitempty"`
	Dy         *float64 `json:"dy,omitempty"`
	Button     *string  `json:"button,omitempty"`
	Action     *string  `json:"action,omitempty"`
	ClickCount *int     `json:"click_count,omitempty"`
	Key        *string  `json:"key,omitempty"`
	Text       *string  `json:"text,omitempty"`
	Modifiers  []string `json:"modifiers,omitempty"`
}

type InputStatusBody struct {
	Active        bool    `json:"active"`
	Reason        *string `json:"reason,omitempty"`
	UDPPort       *uint16 `json:"udp_port,omitempty"`
	DatagramToken *string `json:"datagram_token,omitempty"`
	SecureInput   *bool   `json:"secure_input,omitempty"`
}

type InputAttachBody struct {
	Token string `json:"token"`
}

type MediaControlBody struct {
	Action string   `json:"action"`
	Value  *float64 `json:"value,omitempty"`
}

// --- Phase 3: screen sharing. ---

type ScreenRequestBody struct {
	MaxWidth  *int     `json:"max_width,omitempty"`
	MaxHeight *int     `json:"max_height,omitempty"`
	MaxFps    *int     `json:"max_fps,omitempty"`
	Codecs    []string `json:"codecs"`
}

type ScreenOfferBody struct {
	ScreenSessionID string `json:"screen_session_id"`
	WireSessionID   uint16 `json:"wire_session_id"`
	Codec           string `json:"codec"`
	Width           int    `json:"width"`
	Height          int    `json:"height"`
	Fps             int    `json:"fps"`
	CaptureKind     string `json:"capture_kind"`
	SourceName      string `json:"source_name"`
	BulkToken       string `json:"bulk_token"`
}

type ScreenRejectBody struct {
	Reason string `json:"reason"`
}

type ScreenAttachBody struct {
	ScreenSessionID string `json:"screen_session_id"`
	BulkToken       string `json:"bulk_token"`
}

type ScreenAckBody struct {
	ScreenSessionID string `json:"screen_session_id"`
	AckedSeq        uint32 `json:"acked_seq"`
	RequestKeyframe bool   `json:"request_keyframe"`
}

type ScreenEndBody struct {
	ScreenSessionID string  `json:"screen_session_id"`
	Reason          *string `json:"reason,omitempty"`
}

// Message pairs a decoded body with its type.
type Message struct {
	Type MessageType
	Body interface{}
}

// EncodeMessage produces the canonical JSON of a full envelope for the message.
func EncodeMessage(sessionID string, seq uint64, m Message) ([]byte, error) {
	payload, err := Canonical(m.Body)
	if err != nil {
		return nil, err
	}
	env := Envelope{
		Payload:   json.RawMessage(payload),
		Seq:       seq,
		SessionID: sessionID,
		Type:      string(m.Type),
		Version:   Version,
	}
	return Canonical(env)
}

// DecodeEnvelope splits the header from the payload without interpreting the body.
func DecodeEnvelope(data []byte) (Envelope, error) {
	var env Envelope
	if err := json.Unmarshal(data, &env); err != nil {
		return Envelope{}, fmt.Errorf("not JSON: %w", err)
	}
	return env, nil
}

// DecodeMessage decodes into a typed body. Unknown types return a Message with
// a nil Body and the raw type, mirroring the Swift .unknown case (never fatal).
func DecodeMessage(data []byte) (Envelope, Message, error) {
	env, err := DecodeEnvelope(data)
	if err != nil {
		return Envelope{}, Message{}, err
	}
	body, err := decodeBody(MessageType(env.Type), env.Payload)
	if err != nil {
		return env, Message{}, err
	}
	return env, Message{Type: MessageType(env.Type), Body: body}, nil
}

func decodeBody(t MessageType, payload []byte) (interface{}, error) {
	switch t {
	case TypeHello, TypeHelloAck:
		var b HelloBody
		return unmarshalBody(payload, &b)
	case TypePing, TypePong:
		var b PingBody
		return unmarshalBody(payload, &b)
	case TypeClipboardPush:
		var b ClipboardPushBody
		return unmarshalBody(payload, &b)
	case TypeFileOffer:
		var b FileOfferBody
		return unmarshalBody(payload, &b)
	case TypeFileAccept:
		var b FileAcceptBody
		return unmarshalBody(payload, &b)
	case TypeFileReject:
		var b FileRejectBody
		return unmarshalBody(payload, &b)
	case TypeFileAck:
		var b FileAckBody
		return unmarshalBody(payload, &b)
	case TypePairRequest, TypePairResponse:
		var b PairBody
		return unmarshalBody(payload, &b)
	case TypePairConfirm:
		return struct{}{}, nil
	case TypePairReject:
		var b PairRejectBody
		return unmarshalBody(payload, &b)
	case TypeBulkAttach:
		var b BulkAttachBody
		return unmarshalBody(payload, &b)
	case TypeInputRequest:
		return struct{}{}, nil
	case TypeInputStatus:
		var b InputStatusBody
		return unmarshalBody(payload, &b)
	case TypeInputEvent:
		var b InputEventBody
		return unmarshalBody(payload, &b)
	case TypeInputAttach:
		var b InputAttachBody
		return unmarshalBody(payload, &b)
	case TypeMediaControl:
		var b MediaControlBody
		return unmarshalBody(payload, &b)
	case TypeScreenRequest:
		var b ScreenRequestBody
		return unmarshalBody(payload, &b)
	case TypeScreenOffer:
		var b ScreenOfferBody
		return unmarshalBody(payload, &b)
	case TypeScreenReject:
		var b ScreenRejectBody
		return unmarshalBody(payload, &b)
	case TypeScreenAttach:
		var b ScreenAttachBody
		return unmarshalBody(payload, &b)
	case TypeScreenAck:
		var b ScreenAckBody
		return unmarshalBody(payload, &b)
	case TypeScreenEnd:
		var b ScreenEndBody
		return unmarshalBody(payload, &b)
	case TypeNotification:
		var b NotificationBody
		return unmarshalBody(payload, &b)
	default:
		// Unknown type: inert, not an error (spec §6 invariant).
		return nil, nil
	}
}

func unmarshalBody[T any](payload []byte, b *T) (interface{}, error) {
	if err := json.Unmarshal(payload, b); err != nil {
		return nil, err
	}
	return *b, nil
}

File: android/core/src/main/kotlin/org/conduit/core/wire/Messages.kt
package org.conduit.core.wire

/** Protocol constants shared with Swift + Go (docs/protocol.md). */
object Proto {
    const val VERSION = "0.2"
    const val SERVICE_TYPE = "_mosis-app._tcp"
    const val DEFAULT_CHUNK_SIZE = 512 * 1024

    // Capability identifiers (feature-flag strings; peers use the intersection).
    const val CAP_FILE = "file"
    const val CAP_CLIPBOARD = "clipboard"
    const val CAP_INPUT_INJECT = "input-inject"
    const val CAP_MEDIA_TARGET = "media-target"
    const val CAP_SCREEN_SOURCE = "screen-source"
    const val CAP_SCREEN_VIEW = "screen-view"
    const val CAP_NOTIFY_SOURCE = "notify-source"
    const val CAP_NOTIFY_SHOW = "notify-show"
}

/** The frozen envelope shape (spec §6): version, type, session_id, seq, payload. */
data class Envelope(val version: String, val type: String, val sessionId: String, val seq: Long)

/**
 * A decoded control message: a type string plus the payload as a [Json] value.
 * Keeping the payload as [Json] (rather than one sealed class per body) keeps
 * the core small; capability code reads the fields it needs. Encoding is fully
 * typed via the builders below, which is what the golden vectors pin.
 */
data class Message(val type: String, val payload: Json)

object MessageType {
    const val HELLO = "HELLO"; const val HELLO_ACK = "HELLO_ACK"
    const val PING = "PING"; const val PONG = "PONG"
    const val CLIPBOARD_PUSH = "CLIPBOARD_PUSH"
    const val FILE_OFFER = "FILE_OFFER"; const val FILE_ACCEPT = "FILE_ACCEPT"
    const val FILE_REJECT = "FILE_REJECT"; const val FILE_ACK = "FILE_ACK"
    const val PAIR_REQUEST = "PAIR_REQUEST"; const val PAIR_RESPONSE = "PAIR_RESPONSE"
    const val PAIR_CONFIRM = "PAIR_CONFIRM"; const val PAIR_REJECT = "PAIR_REJECT"
    const val BULK_ATTACH = "BULK_ATTACH"
    const val INPUT_REQUEST = "INPUT_REQUEST"; const val INPUT_STATUS = "INPUT_STATUS"
    const val INPUT_EVENT = "INPUT_EVENT"; const val INPUT_ATTACH = "INPUT_ATTACH"
    const val MEDIA_CONTROL = "MEDIA_CONTROL"
    const val SCREEN_REQUEST = "SCREEN_REQUEST"; const val SCREEN_OFFER = "SCREEN_OFFER"
    const val SCREEN_REJECT = "SCREEN_REJECT"; const val SCREEN_ATTACH = "SCREEN_ATTACH"
    const val SCREEN_ACK = "SCREEN_ACK"; const val SCREEN_END = "SCREEN_END"
    const val NOTIFICATION = "NOTIFICATION"
    // Phase 7: device state + social permissions.
    const val DEVICE_STATE = "DEVICE_STATE"
    const val PERMISSION_REQUEST = "PERMISSION_REQUEST"
    const val PERMISSION_GRANT = "PERMISSION_GRANT"
    const val PERMISSION_REVOKE = "PERMISSION_REVOKE"
}

/** Encodes/decodes control-message envelopes to/from canonical JSON. */
object MessageCodec {
    fun encode(sessionId: String, seq: Long, type: String, payload: Json): ByteArray {
        val env = Json.obj(
            "version" to Json.of(Proto.VERSION),
            "type" to Json.of(type),
            "session_id" to Json.of(sessionId),
            "seq" to Json.of(seq),
            "payload" to payload,
        )
        return CanonicalJson.encode(env)
    }

    /** Re-encodes a payload already in [Json] form (used by conformance to
     *  prove decode→encode is byte-stable). */
    fun encodeMessage(sessionId: String, seq: Long, msg: Message): ByteArray =
        encode(sessionId, seq, msg.type, msg.payload)

    fun decode(data: ByteArray): Pair<Envelope, Message> {
        val root = JsonParser.parse(data).asObj()
        val env = Envelope(
            version = root.getValue("version").str(),
            type = root.getValue("type").str(),
            sessionId = root.getValue("session_id").str(),
            seq = root.getValue("seq").long(),
        )
        return env to Message(env.type, root.getValue("payload"))
    }
}

// ---- Typed payload builders (the wire shapes the golden vectors pin) ----

object Bodies {
    fun hello(
        identity: String, name: String, deviceClass: String, appVersion: String,
        pubkey: ByteArray, capabilities: List<String>, platformWalls: List<String>, listenPort: Int?,
    ): Json {
        val m = linkedMapOf<String, Json>(
            "identity" to Json.of(identity),
            "name" to Json.of(name),
            "device_class" to Json.of(deviceClass),
            "app_version" to Json.of(appVersion),
            "pubkey" to Json.bytes(pubkey),
            "capabilities" to Json.Arr(capabilities.map { Json.of(it) }),
            "platform_walls" to Json.Arr(platformWalls.map { Json.of(it) }),
        )
        if (listenPort != null) m["listen_port"] = Json.of(listenPort)
        return Json.Obj(m)
    }

    fun ping(nonce: String, t: Long): Json = Json.obj("nonce" to Json.of(nonce), "t" to Json.of(t))

    fun clipboard(mime: String, data: ByteArray): Json =
        Json.obj("mime" to Json.of(mime), "data" to Json.bytes(data))

    fun clipboardText(text: String): Json = clipboard("text/plain;charset=utf-8", text.toByteArray(Charsets.UTF_8))

    fun fileOffer(fileId: String, name: String, size: Long, mime: String, sha256: String, chunkSize: Int, chunkCount: Long): Json =
        Json.obj(
            "file_id" to Json.of(fileId), "name" to Json.of(name), "size" to Json.of(size),
            "mime" to Json.of(mime), "sha256" to Json.of(sha256),
            "chunk_size" to Json.of(chunkSize), "chunk_count" to Json.of(chunkCount),
        )

    fun fileAccept(fileId: String, resumeFromChunk: Long, bulkToken: String): Json =
        Json.obj("file_id" to Json.of(fileId), "resume_from_chunk" to Json.of(resumeFromChunk), "bulk_token" to Json.of(bulkToken))

    fun fileReject(fileId: String, reason: String): Json =
        Json.obj("file_id" to Json.of(fileId), "reason" to Json.of(reason))

    fun fileAck(fileId: String, status: String, ackedThrough: Long, message: String? = null): Json {
        val m = linkedMapOf<String, Json>(
            "file_id" to Json.of(fileId), "status" to Json.of(status), "acked_through" to Json.of(ackedThrough),
        )
        if (message != null) m["message"] = Json.of(message)
        return Json.Obj(m)
    }

    fun pair(identity: String, name: String, deviceClass: String, pubkey: ByteArray, tlsHashHex: String, bindingSig: ByteArray): Json =
        Json.obj(
            "identity" to Json.of(identity), "name" to Json.of(name), "device_class" to Json.of(deviceClass),
            "pubkey" to Json.bytes(pubkey), "tls_pubkey_sha256" to Json.of(tlsHashHex), "binding_sig" to Json.bytes(bindingSig),
        )

    fun pairReject(reason: String): Json = Json.obj("reason" to Json.of(reason))
    fun empty(): Json = Json.Obj(emptyMap())
    fun bulkAttach(fileId: String, bulkToken: String): Json =
        Json.obj("file_id" to Json.of(fileId), "bulk_token" to Json.of(bulkToken))

    fun inputEventMove(dx: Double, dy: Double): Json =
        Json.obj("kind" to Json.of("move"), "dx" to Json.Num(numLit(dx)), "dy" to Json.Num(numLit(dy)))

    fun mediaControl(action: String, value: Double? = null): Json {
        val m = linkedMapOf<String, Json>("action" to Json.of(action))
        if (value != null) m["value"] = Json.Num(numLit(value))
        return Json.Obj(m)
    }

    fun notification(appName: String, title: String, body: String, id: String, actions: List<String>? = null): Json {
        val m = linkedMapOf<String, Json>(
            "app_name" to Json.of(appName), "title" to Json.of(title), "body" to Json.of(body), "id" to Json.of(id),
        )
        if (actions != null) m["actions"] = Json.Arr(actions.map { Json.of(it) })
        return Json.Obj(m)
    }

    /** Formats a double the way Swift/Go JSON does: integral values as integers. */
    private fun numLit(d: Double): String =
        if (d == d.toLong().toDouble()) d.toLong().toString() else d.toString()
}
